addons/api/addon/serializers/application.js

```diff
@@ -74,18 +75,38 @@ export default class ApplicationSerializer extends RESTSerializer {
       delete json[key];
     }
     // If an attribute has a `for` option, it must match the
-    // record's `type`, else the attribute is excluded
+    // record's `type`, otherwise the attribute is excluded
     // from serialization.
-    //note: This doesn't handle secrets yet, we can add support for them if needed.
-    if (
-      options?.for &&
-      options.for !== recordType &&
-      !options.for.includes(recordType)
-    ) {
-      if (options.isNestedAttribute) {
-        delete json?.attributes?.[key];
-      } else {
-        delete json[key];
+
+    if (options?.for) {
+      // Check the typeof the `for` option to determine if it is a plugin or not
+      const isPlugin =
+        typeOf(options.for) === 'object' && options.for?.type === 'plugin';
+      const isNotPlugin =
+        typeOf(options.for) === 'string' || typeOf(options.for) === 'array';
+
+      // Helper function to handle deleting attribute and secret keys
+      const deleteKey = (json) => {
+        if (options.isNestedAttribute) {
+          delete json?.attributes?.[key];
+        } else if (options.isNestedSecret && json?.secrets?.[key]) {
+          delete json?.secrets?.[key];
+          if (Object.keys(json.secrets).length === 0) {
+            delete json.secrets;
+          }
+        } else {
+          delete json[key];
+        }
+      };
+
+      // For plugins, the `for` option is an object with a `name` key
+      // For non-plugins, the `for` option is a string or an array
+      if (
+        (isPlugin && !options.for?.name.includes(compositeType)) ||
+        (isNotPlugin && !options.for?.includes(recordType)) ||
+        (isNotPlugin && !options.for === recordType)
```

Comment:
If you include the check for `!options.for === recordType`, I think you'll need to make it something that checks the type:

```suggestion
        (isNotPlugin && !options.for?.includes(recordType) && typeOf(options.for) === 'array') ||
        (isNotPlugin && !options.for === recordType) && typeOf(options.for) === 'string')
```

otherwise you still run in the same issue as it was originally where a string could be checked for in `!options.for?.includes(recordType))`, e.g. `options.for` is `awsExample` and `recordType` is `aws`.

I also don't think this is that likely of an issue so if you want to just keep the `!options.for?.includes(recordType)` check only and just add a comment to this I'd also be fine with that

Reply:
ah yes, it needs the `typeof` check. thanks for the catch!